### PR TITLE
Fix issue 13428: add substituteInout

### PR DESCRIPTION
**Pull request by thewilsonator**
cc @TurkeyMan

### CODE REVIEW

**Comment by TurkeyMan** [comment]
FWIW, I think the arguments are the wrong way around.
I feel like the first argument should be the thing i'm operating on, and the second argument should be the predicate type (the thing I'm copying the mutability from)... does that feel right to you?

**Comment by thewilsonator** [comment]
Yes, tbf, I just copied almost exactly what Simen wrote.

**Comment by Biotronic** [comment]
FWIW, the argument order was chosen to match that of [`CopyTypeQualifiers`](https://dlang.org/phobos/std_traits#CopyTypeQualifiers).

**Comment by thewilsonator** [comment]
I'm not particularly fussed either way.

**Comment by atilaneves** [comment]
I don't know, I feel like it should be `from, to`.

**Comment by Biotronic** [comment]
I revisited this today, and realized it does not handle mixed qualifiers, like `inout(const(int)*)`. As this is a rather big oversight, here's an improved implementation:

```
template SubstituteInout(FromType, ToType)
{
    alias Modifiers(T) = CopyTypeQualifiers!(FromType, T);
    alias Next(T) =    SubstituteInout!(FromType, T);

         static if (is(ToType U ==          immutable U)) alias SubstituteInout =    immutable            Next!U;
    else static if (is(ToType U == shared inout const U)) alias SubstituteInout = shared const Modifiers!(Next!U);
    else static if (is(ToType U == shared inout       U)) alias SubstituteInout = shared       Modifiers!(Next!U);
    else static if (is(ToType U == shared       const U)) alias SubstituteInout = shared const            Next!U;
    else static if (is(ToType U == shared             U)) alias SubstituteInout = shared                  Next!U;
    else static if (is(ToType U ==        inout const U)) alias SubstituteInout =        const Modifiers!(Next!U);
    else static if (is(ToType U ==        inout       U)) alias SubstituteInout =              Modifiers!(Next!U);
    else static if (is(ToType U ==              const U)) alias SubstituteInout =        const            Next!U;
    else static if (is(ToType U ==       U*            )) alias SubstituteInout =                         Next!U*;
    else static if (is(ToType U ==       U[]           )) alias SubstituteInout =                         Next!U[];
    else static if (is(ToType U ==       U[N], size_t N)) alias SubstituteInout =                         Next!U[N];
    else static if (is(ToType U ==       U[K],        K)) alias SubstituteInout =                         Next!U[Next!K];
    else                                                  alias SubstituteInout = ToType;
}

unittest {
    static assert(is(SubstituteInout!(const(string), int) == int));
    static assert(is(SubstituteInout!(const(string), inout(int)[]) == const(int)[]));
    static assert(is(SubstituteInout!(const(string), inout(int)) == const(int)));
    static assert(is(SubstituteInout!(const(string), inout(int)*[][3][int]) == const(int)*[][3][int]));
    static assert(is(SubstituteInout!(const(string), inout(int)[inout(string)]) == const(int)[const(string)]));
    
    static assert(is(SubstituteInout!(const(int), inout(const(int)*)) == const(int*)));
    static assert(is(SubstituteInout!(int, inout(const(int)*)) == const(int)*));
}
```

**Comment by RazvanN7** [comment]
@thewilsonator any chance of continuing with this PR?

File: std/traits.d [hunk 1]
```diff
@@ -5030,6 +5031,34 @@ enum bool isImplicitlyConvertible(From, To) = is(From : To);
...
+        alias SubstituteInout = SubstituteInout!(FromType, SubType)*;
+    else static if (is(ToType == SubType[], SubType))
+        alias SubstituteInout = SubstituteInout!(FromType, SubType)[];
+    else static if (is(ToType == SubType[n], SubType, size_t n))
+        alias SubstituteInout = SubstituteInout!(FromType, SubType)[n];
+    else static if (is(ToType == SubType[KeyType], SubType, KeyType))
+        alias SubstituteInout = SubstituteInout!(FromType, SubType)[SubstituteInout!(FromType, KeyType)];
+    else
+        alias SubstituteInout = ToType;
+}
```
[the remark this answers is not in the document]
Comment (zopsicle) [reply std/traits.d:5052]: DMD bug regarding inout in associative arrays key types is fixed by #16533.

[the remark this answers is not in the document]
Comment (PetarKirov) [reply std/traits.d:5052]: @zopsicle I agree that relying on the compiler is a better approach. I believe your experience puts you in a good position to create an alternative PR with the approach you described.

[the remark this answers is not in the document]
Comment (thewilsonator) [reply std/traits.d:5052]: DMD PR merged, please rebase

[the remark this answers is not in the document]
Comment (zopsicle) [reply std/traits.d:5052]: @thewilsonator This is your PR.

[the remark this answers is not in the document]
Comment (thewilsonator) [reply std/traits.d:5052]: oh, lol

**Comment by LightBender** [comment]
@jmdavis can you ping @thewilsonator to see if this is something that we might need for Phobos 3? Also @TurkeyMan might still be interested in this.

**Pull request closed without merging**